tracer/src/Datadog.Trace/AppSec/Waf/Context.cs

```diff
@@ -10,170 +10,228 @@
+                result.ReturnCode,
+                result.BlockInfo,
+                result.Data);
+        }
 
-                if (ephemeralArgs is not null)
-                {
-                    // WARNING: Don't use ref here, we need to make a copy because ephemeralArgs is on the heap
-                    pwEphemeralArgsValue = ephemeralArgs.ResultDdwafObject;
-                }
+        return result;
+    }
 
-                // WARNING: DO NOT DISPOSE pwPersistentArgs until the end of this class's lifecycle, i.e in the dispose. Otherwise waf might crash with fatal exception.
-                code = _waf.Run(_contextHandle, persistentAddressData != null ? &pwPersistentArgs : null, ephemeralArgs != null ? &pwEphemeralArgsValue : null, ref retNative, timeoutMicroSeconds);
+    public void Dispose(bool disposing)
+    {
+        lock (_stopwatch)
+        {
+            if (_disposed)
+            {
+                return;
             }
 
-            _stopwatch.Stop();
-            _totalRuntimeOverRuns += retNative.TotalRuntime / 1000;
-            var result = new Result(retNative, code, _totalRuntimeOverRuns, (ulong)(_stopwatch.Elapsed.TotalMilliseconds * 1000), isRasp);
-            _wafLibraryInvoker.ResultFree(ref retNative);
+            _disposed = true;
 
-            if (Log.IsEnabled(LogEventLevel.Debug))
+            // WARNING do not move this above, this should only be disposed in the end of the context's life
+            foreach (var encodeResult in _encodeResults)
             {
-                Log.Debug(
-                    "DDAS-0011-00: AppSec In-App WAF returned: {ReturnCode} {BlockInfo} {Data}",
-                    result.ReturnCode,
-                    result.BlockInfo,
-                    result.Data);
+                encodeResult.Dispose();
             }
 
-            return result;
+            _wafLibraryInvoker.ContextDestroy(_contextHandle);
         }
+    }
 
-        public void Dispose(bool disposing)
-        {
-            lock (_stopwatch)
-            {
-                if (_disposed)
-                {
-                    return;
-                }
-
-                _disposed = true;
-
-                // WARNING do not move this above, this should only be disposed in the end of the context's life
-                foreach (var encodeResult in _encodeResults)
-                {
-                    encodeResult.Dispose();
-                }
-
-                _wafLibraryInvoker.ContextDestroy(_contextHandle);
-            }
-        }
+    public void Dispose()
+    {
+        Dispose(true);
+        GC.SuppressFinalize(this);
+    }
 
-        public void Dispose()
-        {
-            Dispose(true);
-            GC.SuppressFinalize(this);
-        }
+    private record UserEventsState
+    {
+        /// <summary>
+        /// Gets or sets a string for the value and bool for if it came from sdk
+        /// </summary>
+        internal UserRecord Id { get; set; } = new(null, false);
```

Comment:
Is it worth creating "null" versions of the `UserRecord`? Are the types mutable? If not, that could save a bunch of allocations - currently we're allocating four objects every time, and then potentially replacing them later?

Reply:
they're mutable, but changed in [55b9f6e](https://github.com/DataDog/dd-trace-dotnet/pull/6431/commits/55b9f6e4d156d227e9d7f60a63008d94e589036e), they're not initialized until we actually need them,
 thanks!